src/UI/Implementation/Render/AbstractComponentRenderer.php

```diff
@@ -223,16 +227,18 @@ private function addTriggererOnLoadCode(Triggerer $triggerer) {
 	 */
 	abstract protected function getComponentInterfaceName();
 
-	// TODO: We might want to cache this.
+
 	private function getMyComponent() {
 		$class = get_class($this);
+		if (isset(self::$component_storage[$class])) {
+			return self::$component_storage[$class];
+		}
 		$matches = array();
 		// Extract component
 		$re = "%ILIAS\\\\UI\\\\Implementation\\\\Component\\\\(\\w+)\\\\(\\w+)%";
-		if (preg_match($re, $class, $matches) !== 1) {
-			throw new \LogicException(
-				"The Renderer needs to be located in ILIAS\\UI\\Implementation\\Component\\*.");
-		}
-		return $matches[1];
+		preg_match($re, $class, $matches);
+		self::$component_storage[$class] = $matches[1];
+
+		return self::$component_storage[$class];
```

Comment:
This will be used in `getTemplate` and nowhere else. If the regexp doesn't match, this will just return null, which seem to break the (implicit) interface of the function that a string is returned. That will happen for your custom component, right? I currently fail to see, how this is related to the choice of the renderer.

Reply:
It does not help to find the renderer, but it leads to a thrown exception for the custom components when ```getTemplate```is called which is final (whyever... To prevent people to talk to ilTemplate?).

Reply:
We cannot effectively prevent people from using `ilTemplate`, it is in the global namespace... :-(

Instead this is an implementation of "Make correct things simple and incorrect things hard." The "correct" thing here is, not to rely on `ilTemplate` directly, but instead use `getTemplate`, which weakens the dependency on a specific rendering engine and a specific location on the filesystem where the template files are located. To implement that we chose to rely on a convention where templates and renderers for the UI-framework are located in the filesystem, which we put in the rules as well. What you experience as a limitation here is in fact a check, that this convention/rule is respected. Since the class and the Renderers are placed in the internal part of the framework (i.e. "not reachable via things on the top level") they are not intended to be extended by inheritance.

How about that suggestion with an overwriteable `getTemplatePath` to still support your usecase? I still wonder how you retrieve your template file in your custom renderer? Or do you use ilTemplate there?

Reply:
> How about that suggestion with an overwriteable getTemplatePath to still support your usecase?

I think this sounds good. The first suggestion works only if using the default templates with the the default renderers are used and combined in new fashions in the new renderer. Custom templates are hard (if not impossible, did it think it through thow, maybe some nasty hack with strange namespaces of the class or relative path argument to ilTemplate could achieve it) to achieve like that. Just overriding getTemplate is probably the way to go. +1 For that. @chfsx Do you agree? If yes, can you revert the removal of the exception and change the final property of setTemplate to protected?

Reply:
I'm fine with the non final `getTemplatePath`and will change the PR accordingly, thanks a lot.